src/librustc_save_analysis/dump_visitor.rs

```diff
@@ -377,26 +373,26 @@ impl<'l, 'tcx> DumpVisitor<'l, 'tcx> {
         ty_params: &'l ast::Generics,
         body: &'l ast::Block,
     ) {
-        if let Some(fn_data) = self.save_ctxt.get_item_data(item) {
-            down_cast_data!(fn_data, DefData, item.span);
-            self.nest_tables(
-                item.id,
-                |v| v.process_formals(&decl.inputs, &fn_data.qualname),
-            );
-            self.process_generic_params(ty_params, &fn_data.qualname, item.id);
-            let hir_id = self.tcx.hir().node_to_hir_id(item.id);
-            self.dumper.dump_def(&access_from!(self.save_ctxt, item, hir_id), fn_data);
-        }
+        let hir_id = self.tcx.hir().node_to_hir_id(item.id);
+        self.nest_tables(item.id, |v| {
+            if let Some(fn_data) = v.save_ctxt.get_item_data(item) {
+                down_cast_data!(fn_data, DefData, item.span);
+                v.process_formals(&decl.inputs, &fn_data.qualname);
```

Comment:
Aside: `process_formals` probably need to be fixed to work like `process_var_decls` -- it probably works poorly with or-patterns now.

Reply:
Thanks for the heads up! Could you please give a simple minimized test case to help me verify the fix/revamp if we were to update the `process_formals` to account for the or-patterns?

Reply:
I know very little about save-analysis so I don't think I can help out with test cases. Basically, I did some changes in https://github.com/rust-lang/rust/pull/64111/files#diff-426e1c382fc57abf697fdeee18e03dde to handle or-patterns in nested positions but I didn't test them. I mostly fused `process_var_decl` and `process_var_decl_multi`. Ideally, you should fuse `process_formals` with `process_var_decl`s as well and get some good code-reuse out of it.

The new thing you have to account for wrt. or-patterns is that users can write `fn foo((Ok(x) | Err(x)): Result<T, E>) { ... }` and such things.

Reply:
I'm hitting an [unimplemented arm](https://github.com/rust-lang/rust/blob/1870537f2701e5aa47080a879b63a4d6b391553b/src/librustc_mir/hair/pattern/_match.rs#L1363) when trying to use an or-pattern in the argument position (where this SA bug has uncovered) so I'd like to come back after this feature has been more fleshed out, if you don't mind :+1:

Reply:
@Xanewok Oh definitely; the feature is totally WIP -- I would recommend filing a keep-sake issue for now and assigning yourself, myself, and @dlrobertson for now.

Reply:
Yeah, still working on that. When https://github.com/rust-lang/rust/pull/63688 is finished the ICE shouldn't happen any more.